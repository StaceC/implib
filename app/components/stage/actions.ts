import { createAction } from 'redux-actions';
import { StagedTrack } from './model';
import STAGING_TRACK_STATES from './constants/StagingTrackStates';

const STAGE_TRACK = 'STAGE_TODO';
const UNSTAGE_TRACK = 'UNSTAGE_TRACK'

const clearCompleted = createAction<StagedTrack, string>(
  STAGE_TRACK,
  (name: string) => ({ id: "", name, status: STAGING_TRACK_STATES.STAGED })
);

const importTracks = createAction<StagedTrack, StagedTrack>(
  UNSTAGE_TRACK,
  (track: StagedTrack) => track
);

export {
  clearCompleted,
  importTracks
}
