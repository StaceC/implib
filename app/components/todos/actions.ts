import { createAction } from 'redux-actions';
import { Todo, Todos, Track } from './model';
import db from "../../db/models";
import * as yauzl from "yauzl";
import { ZipFile, Entry} from "yauzl";

import fs = require('fs');
const path = require('path');
const uuidv4 = require('uuid/v4');
const env = process.env.NODE_ENV || "development";
const appConfig = require(`${__dirname}/../../config/appConfig.json`)[env];

var SAVE_DIR  = "";

if (env == "production") {
  SAVE_DIR = path.join(
    require("electron").remote.app.getPath("appData"),
    //appConfig.appName,
    appConfig.tracksDirName);
} else {
  SAVE_DIR = path.join(appConfig.tracksDirName);
}

import {
  ADD_TODO,
  DELETE_TODO,
  EDIT_TODO,
  COMPLETE_TODO,
  COMPLETE_ALL,
  CLEAR_COMPLETED,
  GET_TODOS_REQUEST,
  GET_TODOS_SUCCESS,
  GET_TODOS_FAILURE,
  IMPORT_TRACKS_REQUEST,
  IMPORT_TRACKS_SUCCESS,
  IMPORT_TRACKS_FAILURE,
  IMPORT_TRACK_REQUEST,
  IMPORT_TRACK_SUCCESS,
  IMPORT_TRACK_FAILURE,
  IMPORT_FILES,
  IMPORT_FILES_SUCCESS,
  IMPORT_FILES_FAILURE,
  IMPORT_FILE,
  IMPORT_FILE_SUCCESS,
  IMPORT_FILE_FAILURE
} from './constants/ActionTypes';

const addTodo = createAction<Todo, string>(
  ADD_TODO,
  (text: string) => ({ id: "", text, completed: false })
);

const deleteTodo = createAction<Todo, Todo>(
  DELETE_TODO,
  (todo: Todo) => todo
);

const editTodo = createAction<Todo, Todo, string>(
  EDIT_TODO,
  (todo: Todo, newText: string) => ({ ...todo, text: newText })
);

const completeTodo = createAction<Todo, Todo>(
  COMPLETE_TODO,
  (todo: Todo) => todo
);

const completeAll = createAction<void>(
  COMPLETE_ALL,
  () => { }
);

const clearCompleted = createAction<void>(
  CLEAR_COMPLETED,
  () => { }
);


// GET TRACKS - DROPZONE FILE GRABBER
export function getTodos() {
  return (dispatch: Function) => {
    dispatch({ type: GET_TODOS_REQUEST });
    return db.Todo.findAll({ raw: true })
      .then((todos) => dispatch(
        getTodosSuccess(todos as Todo[])
      ))
      .catch((error: any) => dispatch( { type: GET_TODOS_FAILURE, payload: null}))
  }
}
const getTodosSuccess = createAction<Todos, Todos>(
  GET_TODOS_SUCCESS,
  (todos: Todos) => todos
);
const getTodosFailure = createAction<void, Todo>(
  GET_TODOS_FAILURE,
  () => {}
  /*
  () => <Todo>{
          text: 'Use Redux with TypeScript',
          completed: false,
          id: "123"
        }
      */
  );


// IMPORT A LIST OF TRACKS -> CALLS IMPORT TRACK
export function importTracks(todos: Todos) {
  return (dispatch: Function) => {
    dispatch({ type: IMPORT_TRACKS_REQUEST });

    const dispatchedTodos = todos.map( track => { return dispatch(importTrack(track)) } );

    return Promise.all(dispatchedTodos)
    .then(() => {
      return dispatch(importTracksSuccess(todos));
    });

  }
}
const importTracksSuccess = createAction<Todos, Todos>(
  IMPORT_TRACKS_SUCCESS,
  (todos: Todos) => todos
);
const importTracksFailure = createAction<void, Todo>(
  IMPORT_TRACKS_FAILURE,
  () => <Todo>{
          text: 'Failed to import tracks',
          completed: false,
          id: "123"
        }
);


// IMPORTING INDIVIDUAL TRACKS
export function importTrack(todo: Todo) {
  return (dispatch: Function) => {
    dispatch({ type: IMPORT_TRACK_REQUEST });
    console.log("Importing Track ["+ todo + "]");
    db.Todo.build( {id: todo.id, text: todo.text, completed: true} )
    .save()
    .then(savedTrack => dispatch(importTrackSuccess(savedTrack as Todo)))
    .catch(error => {
      console.log(error);
      dispatch(importTrackFailure(todo));
    })
  }
}
const importTrackSuccess = createAction<Todo, Todo>(
  IMPORT_TRACK_SUCCESS,
  (todo: Todo) => todo
);
const importTrackFailure = createAction<void, Todo>(
  IMPORT_TRACK_FAILURE,
  () => <Todo>{
          text: 'Failed to import tracks',
          completed: false,
          id: "123"
        }
);

export function importFile(file: File) {
  return (dispatch: Function) => {
    dispatch({ type: IMPORT_FILE });
    unpackArchive(file)
    .then((track: Track) => {
      console.log("Track Name[" + track.name+ "]");
      const newId = uuidv4();
      db.Todo.build( {id: newId.toString(), text: track.name, completed: true} )
      .save()
      .then(savedTrack => {
        console.log(savedTrack);
        dispatch(importFileSuccess(savedTrack.get({plain: true}) as Todo))
      })
      .catch(error => {
        console.log(error);
        dispatch(importFileFailure({id: newId.toString(), text: track.name, completed: false}));
      });
    });
  }
}
const importFileSuccess = createAction<Todo, Todo>(
  IMPORT_FILE_SUCCESS,
  (todo: Todo) => todo
);
const importFileFailure = createAction<void, Todo>(
  IMPORT_FILE_FAILURE,
  () => <Todo>{
          text: 'Failed to import file',
          completed: false,
          id: "123"
        }
);


// IMPORTING INDIVIDUAL TRACKS
export function importFiles(files: File[]) {
  return (dispatch: Function) => {
    dispatch({ type: IMPORT_FILES });
    const dispatchedImportFiles = files.map( file => { return dispatch(importFile(file)) } );

    return Promise.all(dispatchedImportFiles)
    .then(() => {
      return dispatch(importFilesSuccess(files));
    });
  }
}
const importFilesSuccess = createAction<File[], File[]>(
  IMPORT_FILES_SUCCESS,
  (files: File[]) => files
);
const importFilesFailure = createAction<void, Todo>(
  IMPORT_FILES_FAILURE,
  () => <Todo>{
          text: 'Failed to import files',
          completed: false,
          id: "123"
        }
);

function unpackArchive(file: File): Promise<Track> {
  return new Promise(function (resolve,reject) {
    var track: Track = <Track>{};

    yauzl.open((file as any).path, {autoClose: true, lazyEntries: true}, function(err:any, zipfile: ZipFile) {
      if (err) throw err;
      zipfile.readEntry();
      zipfile.on("entry", function(entry: Entry) {
        if (/\/$/.test(entry.fileName)) {
          // Directory file names end with '/'.
          // Note that entires for directories themselves are optional.
          // An entry's fileName implicitly requires its parent directories to exist.
          console.log("Zip Directory: " + entry.fileName);
          const dirname = entry.fileName.replace(/\/$/, "");
          track.name = dirname;
          console.log("Dirname[" + dirname + "]");
          fs.mkdir(path.join(SAVE_DIR, dirname), function(e: any){
            if(!e || (e && e.code === 'EEXIST')){
              console.log("Created directory [" + dirname + "]");
              //do something with contents
            } else {
              //debug
              console.log("Unable to create directory [" + dirname + "]" + e);
            }
          });

        } else {
          // file entry
          zipfile.openReadStream(entry, function(err, readStream) {
            if (err) throw err;
            if(readStream) {
              readStream.on("end", function() {
                zipfile.readEntry();
              });
              console.log("Zip FileName: " + entry.fileName);
              //var filename = entry.fileName.replace(/^.*[\\\/]/, '')
              //console.log("Filename: [" + filename + "]");
              readStream.pipe(fs.createWriteStream(path.join(SAVE_DIR, entry.fileName)));
            }
          });
        }
        // Check if the zip file is still open and read the next entry
        if(zipfile.isOpen) {
          zipfile.readEntry();
        }

        resolve(track);
      });
    });
  })
}

export {
  addTodo,
  deleteTodo,
  editTodo,
  completeTodo,
  completeAll,
  clearCompleted,
  getTodosSuccess,
  getTodosFailure,
  importTracksSuccess,
  importTracksFailure,
  importTrackSuccess,
  importTrackFailure,
  importFilesSuccess,
  importFilesFailure,
  importFileSuccess,
  importFileFailure,
}
