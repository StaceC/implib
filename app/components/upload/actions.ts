import { createAction } from 'redux-actions';

const STAGE_TRACKS_REQUEST= 'STAGE_TRACKS_REQUEST';
const STAGE_TRACKS_SUCCESS = 'STAGE_TRACKS_SUCCEED';
const STAGE_TRACKS_FAILED = 'STAGE_TRACKS_FAILED';
export const STAGE_TRACK = 'STAGE_TRACK';

export function stageTracks(trackFiles: File[]) {
  return (dispatch: Function) => {
    dispatch({ type: STAGE_TRACKS_REQUEST });

    const dispatchedStagedTracks =
      trackFiles.map( file => { return dispatch(stageTrack(file)) } );

    return Promise.all(dispatchedStagedTracks)
    .then(() => {
      return dispatch(stageTracksSuccess());
    })
    .catch((err: Error) => {
      return dispatch(stageTracksFailure(err));
    });

  }
}

const stageTracksSuccess = createAction<void>(
  STAGE_TRACKS_SUCCESS,
  () => { }
);

const stageTracksFailure = createAction<void, Error>(
  STAGE_TRACKS_FAILED,
  (err: Error) => err
);

const stageTrack = createAction<File, File>(
  STAGE_TRACK,
  (trackFile: File) => trackFile
);

export {
  stageTracksSuccess,
  stageTracksFailure,
  stageTrack
}
