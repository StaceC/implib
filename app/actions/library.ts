import { createAction } from 'redux-actions';
import { StagedTrack } from '../components/stage/model';
import db from "../db";
import TRACK_STATES from '../components/library/constants/TrackStates';
import { Track } from '../components/library/model';
import { unpackArchive } from '../utils/unpack';

export const CLEAR_COMPLETED_IMPORTS = 'CLEAR_COMPLETED_IMPORTS';
export const IMPORT_STAGED_TRACKS_REQUEST= 'IMPORT_STAGED_TRACKS_REQUEST';
export const IMPORT_STAGED_TRACKS_SUCCESS = 'IMPORT_STAGED_TRACKS_SUCCEED';
export const IMPORT_STAGED_TRACKS_FAILED = 'IMPORT_STAGED_TRACKS_FAILED';
export const IMPORT_STAGED_TRACK_REQUEST = 'IMPORT_STAGED_TRACK';
export const IMPORT_STAGED_TRACK_SUCCESS = 'IMPORT_STAGED_TRACK_SUCCEED';
export const IMPORT_STAGED_TRACK_FAILED = 'IMPORT_STAGED_TRACK_FAILED';

const clearCompleted = createAction<void, string>(
  CLEAR_COMPLETED_IMPORTS,
  (name: string) => {}
);

export function importStagedTracks(stagedTracks: StagedTrack[]) {
  return (dispatch: Function) => {
    dispatch({ type: IMPORT_STAGED_TRACKS_REQUEST });

    const dispatchedImportTracks =
      stagedTracks.map( track => { return dispatch(importStagedTrack(track)) } );

    return Promise.all(dispatchedImportTracks)
    .then(() => {
      return dispatch(importStagedTracksSuccess());
    })
    .catch((error: Error) => {
      return dispatch(importStagedTracksFailure(error));
    });
  }
}

const importStagedTracksSuccess = createAction<void>(
  IMPORT_STAGED_TRACKS_SUCCESS,
  () => { }
);

const importStagedTracksFailure = createAction<void, Error>(
  IMPORT_STAGED_TRACKS_FAILED,
  (err: Error) => err
);




// IMPORTING INDIVIDUAL TRACKS
export function importStagedTrack(stagedTrack: StagedTrack) {
  return (dispatch: Function) => {
    dispatch({ type: IMPORT_STAGED_TRACK_REQUEST });
    console.log("Importing Track ["+ stagedTrack + "]");
    if(stagedTrack.file) {
      unpackArchive(stagedTrack.file)
      .then(newTrack => {
          db.Track.build( {id: stagedTrack.id, name: newTrack.name, status: TRACK_STATES.IMPORTED } )
          .save()
          .then(savedTrack => dispatch(importStagedTrackSuccess(savedTrack.get({plain: true}) as Track)))
          .catch(error => {
            console.log(error);
            dispatch(importStagedTrackFailure(error));
          });
        }
      ).catch(error => {
        console.log(error);
        dispatch(importStagedTrackFailure(error));
      });
    } else {
      dispatch(importStagedTrackFailure(new Error("Staged Track file attribute is undefined")));
    }
  }
}
const importStagedTrackSuccess = createAction<Track, Track>(
  IMPORT_STAGED_TRACK_SUCCESS,
  (newTrack: Track) => newTrack
);
const importStagedTrackFailure = createAction<void, Error>(
  IMPORT_STAGED_TRACK_FAILED,
  (err: Error) => err
);



export {
  clearCompleted,
  importStagedTracksSuccess,
  importStagedTracksFailure,
  importStagedTrackSuccess,
  importStagedTrackFailure
}
