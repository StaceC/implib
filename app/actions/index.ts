import * as StagingActions from './staging';
export { StagingActions };
import * as LibraryActions from './library';
export { LibraryActions };
export { STAGE_TRACK } from './staging';
