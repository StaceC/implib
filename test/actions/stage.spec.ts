import { spy } from 'sinon';
import {} from 'jest';
import * as actions  from '../../app/actions/staging';
import { testZipFilesArray } from '../resources';

describe('Stage actions', () => {
  it('should bring an array of files into the staging area', () => {
    const fn = actions.stageTracks(testZipFilesArray);
    expect(fn).toBeInstanceOf(Function);
    const dispatch = spy();
    fn(dispatch);
    expect(dispatch.calledWith({ type: actions.stageTrack(testZipFilesArray[0]).type }));
    expect(dispatch.calledWith({ type: actions.stageTracksSuccess().type }));
    expect(dispatch.getCalls().length === 6).toBe(true);
  });

  it('should bring an individual file into the stating area', () => {
    expect(actions.stageTrack(testZipFilesArray[0]).type).toMatchSnapshot();
  });

});
