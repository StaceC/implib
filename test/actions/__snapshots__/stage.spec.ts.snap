// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stage actions should bring an individual file into the stating area 1`] = `"STAGE_TRACK"`;
